### PR TITLE
add mass preserving filter

### PR DESCRIPTION
**Pull request by sandreza**
This PR adds a cutoff filter that preserves the cell average for curved elements. The current filters in CLIMA do not have this property

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by christophernhill** [comment]
@sandreza this looks nice. 

Slightly off topic, but could this affect setups with stretched vertical but otherwise cartesian? I don't think it should, but I had some weirdness with over integration filtering on a stretched grid a while back that I never figured out left on a todo list somewhere......!

**Comment by sandreza** [comment]
> @sandreza this looks nice.
> 
> Slightly off topic, but could this affect setups with stretched vertical but otherwise cartesian? I don't think it should, but I had some weirdness with over integration filtering on a stretched grid a while back that I never figured out, and left on a todo list somewhere......!

I think it depends a bit on how the stretching is done. As long as the Jacobian within an element is constant, the regular filters should be okay. In particular if one is in a cartesian domain, and one just changes the spacing of the elements to be non-uniform, everything should be dandy

**Comment by blallen** [comment]
@sandreza where's the test???

also @jkozdon how should we add this to the main filters module?

**Comment by sandreza** [comment]
> @sandreza where's the test???
> 
> also @jkozdon how should we add this to the main filters module?

This is the test: https://github.com/CliMA/ClimateMachine.jl/blob/ans/sphere/test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter_test.jl
but unfortunately it only works on one MPI rank...

**Comment by blallen** [comment]
can you copy it over and add it to build-kite.yml so it can run on bors in this PR?

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl
Comment (jkozdon) [inline test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl:281]:
```suggestion
        # work around for not being able to `Vars` `@private` arrays
        @unroll for s in 1:nstates
            l_Q2[s] = l_Q[s]
        end
```

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl
Comment (jkozdon) [inline test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl:213]:
```suggestion
         # Load mass matrix and pre-filtered mass weighted quantities to shared memory
```

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl
Comment (jkozdon) [inline test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl:289]:
```suggestion
        # Store result and post-filtered mass weighted quantities
```

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl
Comment (jkozdon) [inline test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl:312]:
```suggestion
        # Reset the element average and store result
        @unroll for s in 1:nstates
```

**Review by jkozdon — CHANGES_REQUESTED** [review]
Nothing really jumped out as me as bad.

Main thing is incorporating a test.

**Comment by jkozdon** [comment]
> also @jkozdon how should we add this to the main filters module?

You should just be able to drop it in to the filters module. I don't see any difficulty with that.

**Comment by jkozdon** [comment]
In principle this should work with any filter matrix I think, so it seems you could actually just add a field to the struct for whether you should use `kernel_apply_filter!` or `kernel_apply_mp_filter!`, and remove the entire `AbstractMassPreservingSpectralFilter`

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl
Comment (sandreza) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl:8]: This one allows for the identity matrix to be applied in a particular direction, if, for example, one wants to use overintegration in the horizontal direction, but not the vertical

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl
Comment (sandreza) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl:187]: > Should we use a different name that `l_Q2` and `l_Qfiltered2`? Maybe `u_` for `@uniform`?

Good suggestion! I'll make the change

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl
Comment (jkozdon) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl:8]: Make sense. I wonder if we want to generally allow this in some way? Or doesn't the filter direction already handle this?

Comment (sandreza) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl:8]: I think that the filter direction might handle this already, but I found the current implementation to be a bit easier to wrap my head around

**Comment by sandreza** [comment]
> In principle this should work with any filter matrix I think, so it seems you could actually just add a field to the struct for whether you should use `kernel_apply_filter!` or `kernel_apply_mp_filter!`, and remove the entire `AbstractMassPreservingSpectralFilter`

I have indeed removed this abstraction, but I wasn't sure about making a change to the other filters in this PR. For now I've just decided to add functionality, leaving a modification to existing functionality for another PR

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl
Comment (sandreza) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/mp_filter.jl]: I have since removed this abstract type!

**Comment by sandreza** [comment]
> @sandreza where's the test???
> 
> also @jkozdon how should we add this to the main filters module?

I have added the test where all the other filters are tested now. It is more of a regression test to show that it replicated the previous behavior. I do not know how to write a test to check to see that it preserves the cell-average (and this is due to mean not knowing how to write a test that takes into account MPI). I have checked that it preserves cell averages on one MPI rank and on the GPU, allowing for discrete conservation

**Comment by blallen** [comment]
bors try

**Comment by blallen** [comment]
> > @sandreza where's the test???
> > also @jkozdon how should we add this to the main filters module?
> 
> I have added the test where all the other filters are tested now. It is more of a regression test to show that it replicated the previous behavior. I do not know how to write a test to check to see that it preserves the cell-average (and this is due to mean not knowing how to write a test that takes into account MPI). I have checked that it preserves cell averages on one MPI rank and on the GPU, allowing for discrete conservation

the test looks good to me. 

@jkozdon when checking `.buildkite/pipeline.yml`, I saw that we only run the filter tests on the CPU. Is that intentional? should we also check them on the GPU?

**Review by blallen — APPROVED** [review]
tests succeeded, so just waiting on approval from @jkozdon now

File: src/Numerics/Mesh/Filters.jl
Comment (jkozdon) [inline src/Numerics/Mesh/Filters.jl:158]:
```suggestion
function modified_filter_matrix(r, Nc, σ)
    N = length(r) - 1
    T = eltype(r)

    @assert N >= 0
    @assert 0 <= Nc
    
    Nc > N && return Array{T}(I, N+1, N+1)

    a, b = GaussQuadrature.legendre_coefs(T, N)
    V = (N == 0 ? ones(T, 1, 1) : GaussQuadrature.orthonormal_poly(r, a, b))

    Σ = ones(T, N + 1)
    Σ[(Nc:N) .+ 1] .= σ.(((Nc:N) .- Nc) ./ (N - Nc))

    V * Diagonal(Σ) / V
end
```
Should we just force the identity matrix?

**Review by jkozdon — COMMENTED** [review]
So I was wondering if you could do an integration test on the sphere using the [weighted sum](https://github.com/CliMA/ClimateMachine.jl/blob/65e5375eaec0505e7a60f30cb61c6e9e31f84a86/src/Arrays/MPIStateArrays.jl#L646-L654) functionality to show that this is needed.

- Compute a solution and compute the integral using weighted 
- Filter without mass preserving and show no mass conservation
- Filter with mass preserving and show it is mass preserving

File: src/Numerics/Mesh/Filters.jl
Comment (sandreza) [reply src/Numerics/Mesh/Filters.jl:158]: I like this suggestion!

**Comment by sandreza** [comment]
> So I was wondering if you could do an integration test on the sphere using the [weighted sum](https://github.com/CliMA/ClimateMachine.jl/blob/65e5375eaec0505e7a60f30cb61c6e9e31f84a86/src/Arrays/MPIStateArrays.jl#L646-L654) functionality to show that this is needed.
> 
> * Compute a solution and compute the integral using weighted
> * Filter without mass preserving and show no mass conservation
> * Filter with mass preserving and show it is mass preserving

This is a great idea! I'll work on adding this test

**Comment by sandreza** [comment]
> So I was wondering if you could do an integration test on the sphere using the [weighted sum](https://github.com/CliMA/ClimateMachine.jl/blob/65e5375eaec0505e7a60f30cb61c6e9e31f84a86/src/Arrays/MPIStateArrays.jl#L646-L654) functionality to show that this is needed.
> 
> * Compute a solution and compute the integral using weighted
> * Filter without mass preserving and show no mass conservation
> * Filter with mass preserving and show it is mass preserving

Just added the test! It's at the bottom of the filters.jl file. I had to re-add the cubedshellwarp since the functionality/name changed

**Review by simonbyrne — APPROVED** [review]
Just need to apply the formatter.

**Review by jkozdon — APPROVED** [review]

**Comment by valeriabarra** [comment]
> Just added the test! It's at the bottom of the filters.jl file. I had to re-add the cubedshellwarp since the functionality/name changed

Hey @sandreza , sorry I am very late in seeing this comment. I'm just curious, why did you need to re-add old code and could have not just imported the function with the current signature as in [this test](https://github.com/CliMA/ClimateMachine.jl/blob/147eee06be26269a596adc31666d82fddfb95ac4/test/Numerics/Mesh/topology.jl#L8)?

**Comment by sandreza** [comment]
@valeriabarra because I was confused and bad with names :) 
luckily @blallen fixed this up in future codes that we wrote

**Comment by valeriabarra** [comment]
Oh, yes, gotcha. I think that's what I've seen in #2171 . All clear then. Thanks!

**Pull request merged**
